Rework Prettier invocation

renchap opened the pull request:
It's usage is no longer recommended as ESlint should not be used anymore for any formatting rules.

More details in this article: https://www.joshuakgoldberg.com/blog/you-probably-dont-need-eslint-config-prettier-or-eslint-plugin-prettier/

This change should speed up ESLint, and formatting issues will no longer show as ESLint warnings when editing, which often happens when developping. It is recommended to configure your editor to format using Prettier on auto-save, otherwise files will be fixed automatically before commit, or when `yarn fix:js` is ran.

The main change here is in the lint/fix scripts, as Prettier needs to be called separately.

I also removed invocations of Prettier from the `lint` and `fix` commands, in favor of new `format` and `format:check` commands. This allows arguments to be passed to the commands, which was not possible when using `&&`. Having one single Prettier invocation is also faster to run, and allows to remove some Github workflows that were nearly instantaneous, but requiring a similar setup time each time.
Prettier is still ran on every file before commit through `lint-staged`, so I do not expect this to change anything for contributors.

renchap committed:
Remove `eslint-plugin-prettier`

It's usage is no longer recommended as ESlint should not be used anymore for any formatting rules.

More details in this article: https://www.joshuakgoldberg.com/blog/you-probably-dont-need-eslint-config-prettier-or-eslint-plugin-prettier/

This change should speed up ESLint, and formatting issues will no longer show as ESLint warnings when editing, which often happens when developping. It is recommended to configure your editor to format using Prettier on auto-save, otherwise files will be fixed automatically before commit, or when `yarn fix:js` is ran.

The main change here is in the lint/fix scripts, as Prettier needs to be called separately.

ThisIsMissEm commented:
@renchap if I'm understanding this correctly, this just moves to using prettier directly and doesn't involve any reformatting of code or anything like that? If so, this wouldn't negatively impact streaming.

The only hold on streaming is anything that modifies the code of the `.js` files, since that'd create a bunch of merge conflicts on PRs just pending review & merge.

nschonni commented on package.json:
```diff
@@ -12,15 +12,15 @@
     "build:development": "cross-env RAILS_ENV=development NODE_ENV=development ./bin/webpack",
     "build:production": "cross-env RAILS_ENV=production NODE_ENV=production ./bin/webpack",
-    "fix:js": "yarn lint:js --fix",
+    "fix:js": "eslint . --ext=.js,.jsx,.ts,.tsx --cache --report-unused-disable-directives --fix && prettier --write \"**/*.{ts,tsx}\" --log-level warn",
```
```suggestion
    "fix:js": "eslint . --ext=.js,.jsx,.ts,.tsx --cache --report-unused-disable-directives --fix && prettier --write \"**/*.{js,jsx,ts,tsx}\" --log-level warn",
```

nschonni commented on package.json:
```diff
@@ -12,15 +12,15 @@
     "i18n:extract": "formatjs extract 'app/javascript/**/*.{js,jsx,ts,tsx}' '--ignore=**/*.d.ts' --out-file app/javascript/mastodon/locales/en.json --format config/formatjs-formatter.js",
     "jest": "cross-env NODE_ENV=test jest",
-    "lint:js": "eslint . --ext=.js,.jsx,.ts,.tsx --cache --report-unused-disable-directives",
+    "lint:js": "eslint . --ext=.js,.jsx,.ts,.tsx --cache --report-unused-disable-directives && prettier --check \"**/*.{ts,tsx}\"",
```
```suggestion
    "lint:js": "eslint . --ext=.js,.jsx,.ts,.tsx --cache --report-unused-disable-directives && prettier --check \"**/*.{js,jsx,ts,tsx}\"",
```

nschonni reviewed (commented):
Although almost all of the JS/JSX is ignored by the `.prettierignore`, it probably makes sense to run it as they should eventually be opted in

renchap commented:
I am wondering if we should drop Prettier from all the lint/fix commands, and use a `format` command that runs Prettier on the whole repo. Files ignored from Prettier would not be changed, and every other supported file will be formatted at once.

Main advantage is that we no longer would have `&&` in scripts, which is not working nice when you want to add options to a script (let's say `yarn fix --verbose`).

Main drawback is that you would need to run 2 commands to fix issues (`yarn fix && yarn format`).

As Prettier is ran automatically on every changed file in git pre-commit, maybe it is not a real issue?

ThisIsMissEm commented:
@renchap lint-staged already calls prettier on every file from what I can tell? But I'd be in favour of a `format:check` and `format:fix`

renchap committed:
Use a dedicaced `format` yarn command for formatting

renchap commented:
Updated, introduced `yarn format` and `yarn format:check` in a second commit. This allows to remove all the scripts and workflow that ran Prettier on specific extensions and were wasteful.

renchap committed:
Add back the format:check GH:A workflow

renchap committed:
Do not use an extension allowlist for Prettier, let it process everything
